docs/python/tutorial-fastapi.md

```diff
@@ -0,0 +1,595 @@
+3. Next, search for "Python inlay type hints", and enable inlay hints for Variable Types and Function Return Types:
+    ![Two Python Analysis Type Hints settings being enabled in the settings UI page: for Function Return Types and for Variable Types.](images/fastapi-tutorial/function_and_variable_return_type_hint_settings.png)
+
+Now let's create a Pydantic model for grocery list items.
+1. Create a new Python file (`kb(workbench.action.files.newFile)` and select "Python File")
+2. Add the following lines to the file, and then save it in the `groceries-plugin` folder as `models.py` (`kb(workbench.action.files.saveAs)`):
+    ```python
+    from typing import Optional
+    from pydantic import BaseModel
+
+    class ItemPayload(BaseModel):
+        item_id: Optional[int]
+        item_name: str
+        quantity: int
+    ```
+
+## Create Routes
+Now we need a place to store the grocery list items. For simplicity, let’s start with an empty dictionary.
+1. First, let's import all the packages we will need for the sample. Open the main.py file and replace the first import line with the following ones:
+    ```python
+    from fastapi import FastAPI, HTTPException
+
+    from models import ItemPayload
+    ```
+2. Now add the following line right below `app = FastAPI()`:
+    ```python
+    grocery_list: dict[int, ItemPayload] = {}
+    ```
+
+    This creates a new empty dictionary that will receive keys of type int (which will be each item id) and values of the ItemPayload type.
+
+We’ll now define routes that will allow us to add and retrieve individual items, as well as list all items in the grocery list.
+
+3. Add the following route at the end of the `main.py` file:
+    ```python
+    # Route to add a item
+    @app.post("/items/{item_name}/{quantity}")
+    def add_item(item_name: str, quantity: int):
+        if quantity <= 0:
+            raise HTTPException(status_code=400, detail="Quantity must be greater than 0.")
+        # if item already exists, we'll just add the quantity.
+        # get all item names
+        items_ids = {item.item_name: item.item_id if item.item_id is not None else 0 for item in grocery_list.values()}
+        if item_name in items_ids.keys():
+            # get index of item_name in item_ids, which is the item_id
+            item_id = items_ids[item_name]
+            grocery_list[item_id].quantity += quantity
+    # otherwise, create a new item
+        else:
+            # generate an id for the item based on the highest ID in the grocery_list
+            item_id = max(grocery_list.keys()) + 1 if grocery_list else 0
+            grocery_list[item_id] = ItemPayload(
+                item_id=item_id, item_name=item_name, quantity=quantity
+            )
+
+    return {"item": grocery_list[item_id]}
+    ```
+
+    You might notice Pylance adds inlay hints with the function return type, as well as the types for `item_ids` and `item_id`. You can double click on each suggestion to insert them into the code:
```

Comment:
When running through this, Pylance didn't add hints for all those in the screenshot, should we consider adding something that says the user can add these in manually as well? I don't want users to think something is wrong with their setup or code because it doesn't match exactly.

Reply:
added some words to make it clearer that inlay hints are optional in https://github.com/microsoft/vscode-docs/commit/5b8cdfcf5f2f647ca92188167b6efdb98f657a4a. But if you ever see this again @cwebster-99 would you mind filing an issue in the Pylance repo? It sounds like a bug